researchapp/services/resources.py

```diff
@@ -0,0 +1,90 @@
+""" Resources Service
+"""
+from researchapp.models import DBSession
+from researchapp.models.resources import Resource
+from researchapp.services import fhir
+
+
+S4S_RESOURCES = {
+    'Smoking status': [
+        'Observation?category=social-history&patient={patientId}',
+    ],
+    'Problems': [
+        'Condition?patient={patientId}',
+    ],
+    'Medications and allergies': [
+        'MedicationOrder?patient={{patientId}}',
+        'MedicationStatement?patient={{patientId}}',
+        'MedicationDispense?patient={{patientId}}',
+        'MedicationAdministration?patient={{patientId}}',
+        'AllergyIntolerance?patient={{patientId}}',
+    ],
+    'Lab results': [
+        'Observation?category=laboratory?patient={patientId}',
+    ],
+    'Vital signs': [
+        'Observation?category=vital-signs?patient={patientId}',
+    ],
+    'Procedures': [
+        'Procedure?patient={patientId}',
+    ],
+    'Immunizations': [
+        'Immunization?patient={patientId}',
+    ],
+    'Patient documents': [
+        'DocumentReference?patient={patientId}',
+    ],
+}
+
+
+def resource_service(which='db'):
+    """ factory method """
+    if which == 'db':
+        return DbService()
+
+
+class DbService(object):
+    """ Database backed ParticipantService
+    """
+
+    def __init__(self):
+        """ init """
+
+    def sync(self, participant, provider):
+        """ Sync, FOR SCIENCE! """
+
+        for ccds, endpoints in S4S_RESOURCES.items():
+            for endpoint in endpoints:
+                try:
+                    bundle = fhir.query(participant, provider, endpoint)
+                except AssertionError as err:
+                    print(err, endpoint)
+                    continue
+
+                if bundle['total'] == 0:
+                    continue
+                for entry in bundle['entry']:
+                    self.save_resource(entry['resource'],
+                                       participant,
+                                       provider)
+
+    def save_resource(self, entry, participant, provider):
+        """ Save a resource to the database. """
+        import json
+
+        resource = Resource(entry=json.dumps(entry),
+                            fhir_id=entry['id'],
```

Comment:
Namespacw issue: Either this id needs to be normalized to an absolute URL including the server base, or (better still, I think) the FHIR baseuURL needs to be captured as a separate field. Otherwise we could see conflict between data from different fhir servers.